jd4/case.py

```diff
@@ -90,70 +98,175 @@ def __init__(self, open_input, open_output, time_ns, memory_bytes, score):
+
+            async def prepare_judge_sandbox():
+                await judge_sandbox.reset()
+                return await judge_package.install(judge_sandbox)
+
+            user_executable, judge_executable = \
+                await gather(prepare_user_sandbox(), prepare_judge_sandbox())
+            user_stdin_file = path.join(user_sandbox.in_dir, 'stdin')
+            mkfifo(user_stdin_file)
+            user_stdout_file = path.join(user_sandbox.in_dir, 'stdout')
+            mkfifo(user_stdout_file)
+            judge_stdin_file = path.join(judge_sandbox.in_dir, 'stdin')
+            link(user_stdout_file, judge_stdin_file)
+            user_stderr_file = path.join(user_sandbox.in_dir, 'stderr')
+            mkfifo(user_stderr_file)
+            judge_stdout_file = path.join(judge_sandbox.in_dir, 'stdout')
+            mkfifo(judge_stdout_file)
+            judge_stderr_file = path.join(judge_sandbox.in_dir, 'stderr')
+            mkfifo(judge_stderr_file)
+            judge_extra_file = path.join(judge_sandbox.in_dir, 'extra')
+            mkfifo(judge_extra_file)
+            with socket(AF_UNIX, SOCK_STREAM | SOCK_NONBLOCK) as user_cgroup_sock, \
+                 socket(AF_UNIX, SOCK_STREAM | SOCK_NONBLOCK) as judge_cgroup_sock:
+                user_cgroup_sock.bind(path.join(user_sandbox.in_dir, 'cgroup'))
+                judge_cgroup_sock.bind(path.join(judge_sandbox.in_dir, 'cgroup'))
+                user_cgroup_sock.listen()
+                judge_cgroup_sock.listen()
+                user_execute_task = loop.create_task(user_executable.execute(
+                    user_sandbox,
+                    stdin_file='/in/stdin',
+                    stdout_file='/in/stdout',
+                    stderr_file='/in/stderr',
+                    cgroup_file='/in/cgroup'))
+                judge_execute_task = loop.create_task(judge_executable.execute(
+                    judge_sandbox,
+                    stdin_file='/in/stdin',
+                    stdout_file='/in/stdout',
+                    stderr_file='/in/stderr',
+                    extra_file='/in/extra',
+                    cgroup_file='/in/cgroup'))
+                others_task = gather(
+                    loop.run_in_executor(None, self.do_input, user_stdin_file),
+                    loop.run_in_executor(None, self.do_input, judge_extra_file),
+                    read_pipe(user_stderr_file, MAX_STDERR_SIZE),
+                    read_pipe(judge_stdout_file, MAX_STDERR_SIZE),
+                    read_pipe(judge_stderr_file, MAX_STDERR_SIZE),
+                    wait_cgroup(user_cgroup_sock,
+                                user_execute_task,
+                                self.time_ns,
+                                self.memory_bytes,
+                                PROCESS_LIMIT),
+                    wait_cgroup(judge_cgroup_sock,
+                                judge_execute_task,
+                                DEFAULT_TIME_NS,
```

Comment:
DEFAULT_TIME_NS is too short for some cases?

Reply:
Good catch, I think we need configurable time and memory limit for the custom judge.

Reply:
I think this configuration is global.